Enable Windows-Arm32 build and packaging for iotedged

chandde opened the pull request:
This PR is to enable win-arm build and packaging, including changes primarily in building scripts (yaml and ps1) to utilize private rust compiler and some private rust crate forks. 

1. Rust Arm compiler
We cannot use the official tools, and we have to use a private version. Some dependencies in LLVM will not be ready in short term. The private tool is uploaded to https://iottools.blob.core.windows.net/iotedge-armtools/rust-windows-arm.zip and is downloaded and extracted at build time
2. Docker and Dockerd
I also uploaded the arm32 version docker and dockerd, https://iottools.blob.core.windows.net/iotedge-armtools/moby-cli.zip, https://iottools.blob.core.windows.net/iotedge-armtools/moby-engine.zip, both will be downloaded and extracted when generating the package
3. Rust Crates dependencies
There are a couple crates that we forked and made private changes so it can build win-arm32. While we upstream all these dependency changes, we patch these crates with the forks, by appending a [patch] section in the root `cargo.toml` in edgelet, also override `cargo.lock` with a pre-configured `cargo.winarm.lock` in the same folder, this lock file has all the crates dependencies pointing to our forks
4. Build scripts
The build scripts have been updated to support arm, e.g. to install arm-windows openssl instead of x64, to download the private rust toolchain instead of the official one, etc.. 
5. Known issues with fixes
When building openssl-sys, it fails to find cl.exe to expand a .c file, I added the folder of the x64-on-x64 CL.exe to %PATH. 
When generating the package, makecat.exe cannot be found by pkggen.exe, so I added folder of makecat.exe to %PATH%

arsing commented:
>also override cargo.lock with a pre-configured cargo.winarm.lock in the same folder, this lock file has all the crates dependencies pointing to our forks

This will become outdated as soon as we change any deps that modify the real `Cargo.lock`. Why do you need this file? `cargo build` will automatically make the minimal change necessary to the real `Cargo.lock` when it detects the `[patch]`es added by the build script.

arsing commented on edgelet/build/windows/build.ps1:
```diff
@@ -1,23 +1,74 @@
+    Write-Host $(Get-Command cl.exe).Path
+}
+
+Write-Host "$cargo build $(if (-Not $Arm) { '--all'} else {'--target thumbv7a-pc-windows-msvc'}) $(if ($Release) { '--release' }) --manifest-path $ManifestPath"
```
Why is `--all` conditional on `$Arm` ? Shouldn't it be

```powershell
"$cargo build --all $(if ($Arm) {'--target thumbv7a-pc-windows-msvc'}) $(if ($Release) { '--release' }) --manifest-path $ManifestPath"
```

?

chandde replied:
I'll make --all always available

chandde replied:
reverted back so --all is always used

arsing commented on edgelet/build/windows/openssl.ps1:
```diff
@@ -2,39 +2,47 @@
     {
         Write-Host "Installing vcpkg from github..."
-        git clone https://github.com/Microsoft/vcpkg $env:HOMEDRIVE\vcpkg
+        git clone https://github.com/Microsoft/vcpkg $env:SystemDrive\vcpkg
```
Why do you need to change this? The script is intended for everyone to use, not just the VSTS agents, and users will probably not have access to write to `C:\vcpkg`

chandde replied:
When I tried on my VSTS server, $env:HOMEDRIVE was not defined, so I had to change to something that works. What's the recommended var to use here? I can change back to HOMEDRIVE and try it on MsAzure pipeline

chandde replied:
Reverted back to $env:HOMEDRIVE

arsing commented on edgelet/hsm-sys/azure-iot-hsm-c/CMakeLists.txt:
```diff
@@ -186,3 +186,8 @@ set (CPACK_RPM_PACKAGE_LICENSE "Proprietary")
 
 include (CPack)
 
+if (${run_unittests})
```
Why do you need this?

chandde replied:
Philip had this change in his iotedge fork so I inherited that to my fork. @philipktlin ?

philipktlin replied:
If I remember correctly, this is to avoid Checkin/CI build fail.  Try remove it and see if Checkin/CI build pass.

chandde replied:
checking...

chandde replied:
revert it for now and will kick off a build to verify

chandde replied:
Without this new section, below error is hit multiple times, SHFileOperation is from Shell32.lib, so looks like we need to keep the change.

```
test_utils.obj : error LNK2019: unresolved external symbol **__imp_SHFileOperationA** referenced in function hsm_test_util_delete_dir [I:\iotedge-1\edgelet\target\thumbv7a-pc-windows-msvc\release\build\hsm-sys-39e8967a78d2a48b\out\build\tests\edge_hsm_util_int\edge_hsm_util_int_exe.vcxproj]
I:\iotedge-1\edgelet\target\thumbv7a-pc-windows-msvc\release\build\hsm-sys-39e8967a78d2a48b\out\build\tests\edge_hsm_util_int\Release\edge_hsm_util_int_exe.exe : fatal error LNK1120: 1 unresolved externals [I:\iotedge-1\edgelet\target\thumbv7a-pc-windows-msvc\release\build\hsm-sys-39e8967a78d2a48b\out\build\tests\edge_hsm_util_int\edge_hsm_util_int_exe.vcxproj]
```

arsing replied:
Oh right, I remember finding this on Philip's machine. We concluded that the amd64 SDK was pulling in shell32.lib implicitly but arm SDK was not.

chandde commented:
> > also override cargo.lock with a pre-configured cargo.winarm.lock in the same folder, this lock file has all the crates dependencies pointing to our forks
> 
> This will become outdated as soon as we change any deps that modify the real `Cargo.lock`. Why do you need this file? `cargo build` will automatically make the minimal change necessary to the real `Cargo.lock` when it detects the `[patch]`es added by the build script.

I removed cargo.winarm.lock, and have cargo to update the dependencies, However I added a few cargo update commands in the arm build to manually update them, please refer to the build.ps1 file. If I let cargo build to update automatically, it'll end up patching only some of the crates instead of all crates. Cargo build will print something as below, if I don't manually run the cargo update commands for ARM build, just FYI...

```
D:\i1\iotedge\edgelet>cargo build --all --target thumbv7a-pc-windows-msvc --release
    Updating git repository `https://github.com/philipktlin/mio-uds-windows`
    Updating git repository `https://github.com/philipktlin/tokio-uds-windows`
    Updating git repository `https://github.com/philipktlin/backtrace-rs`
    Updating git repository `https://github.com/philipktlin/cmake-rs`
    Updating git repository `https://github.com/philipktlin/dtoa`
    Updating git repository `https://github.com/philipktlin/iovec`
    Updating git repository `https://github.com/philipktlin/mio`
    Updating git repository `https://github.com/philipktlin/miow`
    Updating git repository `https://github.com/philipktlin/hjson-rust`
    Updating git repository `https://github.com/philipktlin/winapi-rs`
    Updating git repository `https://github.com/philipktlin/hyperlocal-windows`
warning: Patch `mio-uds-windows v0.1.0 (https://github.com/philipktlin/mio-uds-windows?branch=arm#581fcd3d)` was not used in the crate graph.
Patch `tokio-uds-windows v0.1.0 (https://github.com/philipktlin/tokio-uds-windows?branch=arm#812f4df6)` was not used in the crate graph.
Check that the patched package version and available features are compatible
with the dependency requirements. If the patch has a different version from
what is locked in the Cargo.lock file, run `cargo update` to use the new
version. This may also occur with an optional dependency that is not enabled.
```

chandde commented:
latest build is here https://msazure.visualstudio.com/One/_build/results?buildId=21646906, it has all changes to address the comments, 

1. Change back to $env:HOMEDRIVE
2. Removed cargo.winarm.lock and run "cargo update" manually to patch the crates
3. Made --all available to all builds

arsing commented on edgelet/build/windows/build.ps1:
```diff
@@ -1,23 +1,80 @@
+    Add-Content -Path cargo.toml -Value $ForkedCrates
+
+    Write-Host "Running cargo update to lock the crate forks required by arm build"
+    Invoke-Expression "$cargo update -p https://github.com/Azure/mio-uds-windows.git#mio-uds-windows:0.1.0"
```
I assume you're downgrading mio because it happens to be upgraded to v0.6.16 by the previous `cargo update`, but there is no reason to assume that mio is the only one that needs to be downgraded. If mio got updated to v0.6.17 and that version also pulled in an updated version of some other dependency, then that second dependency would remain at the newer version.

Instead you should only need:

```powershell
Invoke-Expression "$cargo update -p winapi:0.3.5 --precise 0.3.5"
Invoke-Expression "$cargo update -p mio-uds-windows -p tokio-uds-windows"
```

chandde replied:
these two lines aren't enough, the patch list and the four update commands are the minimal I can get.
From my experience, with or without .git when referencing a crate does have impact to cargo update, it'll end up behaving differently.
Also, I need to cargo update to the specific version so the patch can take effect, because Philip's fork for mio is 0.6.14.

arsing replied:
>these two lines aren't enough, the patch list and the four update commands are the minimal I can get.

They're enough. I wrote my comment after I tested it on my machine.

>Also, I need to cargo update to the specific version so the patch can take effect, because Philip's fork for mio is 0.6.14.

The version in our lockfile is also 0.6.14.

chandde replied:
Did you cargo build after you updated the lock files? and on what branch? If I only patch mio-uds-windows, without patching tokio-uds-windows or hyperlocal-windows, after I ran your update commands then tried to build, the public mio-uds-windows was added back, which brought in winapi 0.2.8. the public mio-uds-windows was pulled in by hyperlocal-windows. I thought we're trying to get rid of 0.2.8 winapi...

```
name = "hyperlocal-windows"
version = "0.1.0"
source = "git+https://github.com/Azure/hyperlocal-windows#2bd432bbbfb5b1cf38429733dd9a593c7b97a850"
dependencies = [
 "futures 0.1.24 (registry+https://github.com/rust-lang/crates.io-index)",
 "hex 0.3.2 (registry+https://github.com/rust-lang/crates.io-index)",
 "hyper 0.12.17 (registry+https://github.com/rust-lang/crates.io-index)",
 **"mio-uds-windows 0.1.0 (git+https://github.com/Azure/mio-uds-windows.git)",**

[[package]]
name = "mio-uds-windows"
version = "0.1.0"
source = "git+https://github.com/Azure/mio-uds-windows.git#87a4a9970e668fdbe9205f4039967e175182c70e"
dependencies = [
 "iovec 0.1.2 (git+https://github.com/philipktlin/iovec?branch=arm)",
 "kernel32-sys 0.2.2 (registry+https://github.com/rust-lang/crates.io-index)",
 "lazycell 1.2.0 (registry+https://github.com/rust-lang/crates.io-index)",
 "log 0.4.5 (registry+https://github.com/rust-lang/crates.io-index)",
 "mio 0.6.14 (git+https://github.com/philipktlin/mio?branch=arm)",
 "miow 0.2.1 (registry+https://github.com/rust-lang/crates.io-index)",
 "winapi 0.2.8 (registry+https://github.com/rust-lang/crates.io-index)",
 "ws2_32-sys 0.2.1 (registry+https://github.com/rust-lang/crates.io-index)",
]
```

and here's the cargo build output, see first line

```
I:\iotedge-1\edgelet>cargo build --all --target thumbv7a-pc-windows-msvc --release
    Updating git repository `https://github.com/Azure/mio-uds-windows.git`
    Updating crates.io index
   Compiling winapi v0.2.8
   Compiling cmake v0.1.30 (https://github.com/philipktlin/cmake-rs?branch=arm#886c0dd7)
   Compiling regex-syntax v0.5.6
   Compiling openssl-sys v0.9.36
   Compiling winapi v0.3.5 (https://github.com/philipktlin/winapi-rs?branch=arm/v0.3.5#12a51f5b)
   Compiling serde v1.0.84
   Compiling qu
```
and eventually build failed,

```
error[E0412]: cannot find type `ULONG_PTR` in this scope
  --> C:\Users\dengy\.cargo\registry\src\github.com-1ecc6299db9ec823\winapi-0.2.8\src\basetsd.rs:86:19
   |
86 | pub type SIZE_T = ULONG_PTR;
   |                   ^^^^^^^^^ not found in this scope

error[E0412]: cannot find type `ULONG_PTR` in this scope
  --> C:\Users\dengy\.cargo\registry\src\github.com-1ecc6299db9ec823\winapi-0.2.8\src\basetsd.rs:87:25
   |
87 | pub type PSIZE_T = *mut ULONG_PTR;
   |                         ^^^^^^^^^ not found in this scope

error[E0412]: cannot find type `LONG_PTR` in this scope
  --> C:\Users\dengy\.cargo\registry\src\github.com-1ecc6299db9ec823\winapi-0.2.8\src\basetsd.rs:88:20
   |
88 | pub type SSIZE_T = LONG_PTR;
   |                    ^^^^^^^^ not found in this scope

error[E0412]: cannot find type `LONG_PTR` in this scope
  --> C:\Users\dengy\.cargo\registry\src\github.com-1ecc6299db9ec823\winapi-0.2.8\src\basetsd.rs:89:26
   |
89 | pub type PSSIZE_T = *mut LONG_PTR;
   |                          ^^^^^^^^ not found in this scope

error[E0412]: cannot find type `ULONG_PTR` in this scope
  --> C:\Users\dengy\.cargo\registry\src\github.com-1ecc6299db9ec823\winapi-0.2.8\src\basetsd.rs:90:22
   |
90 | pub type DWORD_PTR = ULONG_PTR;
   |                      ^^^^^^^^^ not found in this scope

error[E0412]: cannot find type `ULONG_PTR` in this scope
  --> C:\Users\dengy\.cargo\registry\src\github.com-1ecc6299db9ec823\winapi-0.2.8\src\basetsd.rs:91:28
   |
91 | pub type PDWORD_PTR = *mut ULONG_PTR;
   |                            ^^^^^^^^^ not found in this scope

error[E0412]: cannot find type `ULONG_PTR` in this scope
  --> C:\Users\dengy\.cargo\registry\src\github.com-1ecc6299db9ec823\winapi-0.2.8\src\basetsd.rs:98:22
   |
98 | pub type KAFFINITY = ULONG_PTR;
   |                      ^^^^^^^^^ not found in this scope

error[E0412]: cannot find type `ULONG_PTR` in the crate root
  --> C:\Users\dengy\.cargo\registry\src\github.com-1ecc6299db9ec823\winapi-0.2.8\src\cfgmgr32.rs:43:28
   |
43 | pub type CONFLICT_LIST = ::ULONG_PTR;
   |                            ^^^^^^^^^ not found in the crate root

error[E0412]: cannot find type `TBBUTTON` in this scope
   --> C:\Users\dengy\.cargo\registry\src\github.com-1ecc6299db9ec823\winapi-0.2.8\src\commctrl.rs:516:27
    |
516 | pub type PTBBUTTON = *mut TBBUTTON;
    |                           ^^^^^^^^ help: a type alias with a similar name exists: `PTBBUTTON`

error[E0412]: cannot find type `TBBUTTON` in this scope
   --> C:\Users\dengy\.cargo\registry\src\github.com-1ecc6299db9ec823\winapi-0.2.8\src\commctrl.rs:517:28
    |
517 | pub type LPTBBUTTON = *mut TBBUTTON;
    |                            ^^^^^^^^ help: a type alias with a similar name exists: `PTBBUTTON`

error[E0412]: cannot find type `TBBUTTON` in this scope
   --> C:\Users\dengy\.cargo\registry\src\github.com-1ecc6299db9ec823\winapi-0.2.8\src\commctrl.rs:518:31
    |
518 | pub type LPCTBBUTTON = *const TBBUTTON;
    |                               ^^^^^^^^ help: a type alias with a similar name exists: `PTBBUTTON`

error[E0412]: cannot find type `UINT_PTR` in the crate root
   --> C:\Users\dengy\.cargo\registry\s
```

arsing replied:
```powershell
$ git diff --patch HEAD~1

diff --git a/edgelet/Cargo.toml b/edgelet/Cargo.toml
index f86c77d6..33a16bb8 100644
--- a/edgelet/Cargo.toml
+++ b/edgelet/Cargo.toml
@@ -29,3 +29,16 @@ members = [
     "tokio-named-pipe",
     "win-logger",
 ]
+
+[patch.crates-io]
+backtrace = { git = "https://github.com/philipktlin/backtrace-rs", branch = "arm" }
+cmake = { git = "https://github.com/philipktlin/cmake-rs", branch = "arm" }
+dtoa = { git = "https://github.com/philipktlin/dtoa", branch = "arm" }
+iovec = { git = "https://github.com/philipktlin/iovec", branch = "arm" }
+mio = { git = "https://github.com/philipktlin/mio", branch = "arm" }
+miow = { git = "https://github.com/philipktlin/miow", branch = "arm" }
+serde-hjson = { git = "https://github.com/philipktlin/hjson-rust", branch = "arm" }
+winapi = { git = "https://github.com/philipktlin/winapi-rs", branch = "arm/v0.3.5" }
+
+[patch."https://github.com/Azure/mio-uds-windows.git"]
+mio-uds-windows = { git = "https://github.com/philipktlin/mio-uds-windows.git", branch = "arm" }
diff --git a/edgelet/build/windows/build.ps1 b/edgelet/build/windows/build.ps1
index 32628cad..d921e565 100644
--- a/edgelet/build/windows/build.ps1
+++ b/edgelet/build/windows/build.ps1
@@ -37,24 +37,16 @@ miow = { git = "https://github.com/philipktlin/miow", branch = "arm" }
 serde-hjson = { git = "https://github.com/philipktlin/hjson-rust", branch = "arm" }
 winapi = { git = "https://github.com/philipktlin/winapi-rs", branch = "arm/v0.3.5" }

-[patch."https://github.com/Azure/mio-uds-windows"]
+[patch."https://github.com/Azure/mio-uds-windows.git"]
 mio-uds-windows = { git = "https://github.com/philipktlin/mio-uds-windows", branch = "arm" }
-
-[patch."https://github.com/Azure/hyperlocal-windows"]
-hyperlocal-windows = { git = "https://github.com/philipktlin/hyperlocal-windows", branch = "arm" }
-
-[patch."https://github.com/Azure/tokio-uds-windows"]
-tokio-uds-windows = { git = "https://github.com/philipktlin/tokio-uds-windows", branch = "arm" }
 "@

     Write-Host "Append cargo.toml with $ForkedCrates"
     Add-Content -Path cargo.toml -Value $ForkedCrates

     Write-Host "Running cargo update to lock the crate forks required by arm build"
-    Invoke-Expression "$cargo update -p https://github.com/Azure/mio-uds-windows.git#mio-uds-windows:0.1.0"
-    Invoke-Expression "$cargo update -p mio --precise 0.6.14"
-    Invoke-Expression "$cargo update -p https://github.com/Azure/tokio-uds-windows.git#tokio-uds-windows:0.1.0"
-    Invoke-Expression "$cargo update -p winapi --precise 0.3.5"
+    Invoke-Expression "$cargo update -p winapi:0.3.5 --precise 0.3.5"
+    Invoke-Expression "$cargo update -p mio-uds-windows"
 }


diff --git a/edgelet/edgelet-http/Cargo.toml b/edgelet/edgelet-http/Cargo.toml
index 312c5239..ff155897 100644
--- a/edgelet/edgelet-http/Cargo.toml
+++ b/edgelet/edgelet-http/Cargo.toml
@@ -40,9 +40,9 @@ tokio-uds = "0.2"
 [target.'cfg(windows)'.dependencies]
 hyper-named-pipe = { path = "../hyper-named-pipe" }
 hyperlocal-windows = { git = "https://github.com/Azure/hyperlocal-windows" }
-mio-uds-windows = { git = "https://github.com/Azure/mio-uds-windows" }
+mio-uds-windows = { git = "https://github.com/Azure/mio-uds-windows.git" }
 tokio-named-pipe = { path = "../tokio-named-pipe" }
-tokio-uds-windows = { git = "https://github.com/Azure/tokio-uds-windows" }
+tokio-uds-windows = { git = "https://github.com/Azure/tokio-uds-windows.git" }
 winapi = { version = "0.3.5", features = ["winsock2"] }

 [dev-dependencies]
diff --git a/edgelet/edgelet-test-utils/Cargo.toml b/edgelet/edgelet-test-utils/Cargo.toml
index 82946a6c..fc6aa003 100644
--- a/edgelet/edgelet-test-utils/Cargo.toml
+++ b/edgelet/edgelet-test-utils/Cargo.toml
@@ -24,5 +24,5 @@ hyperlocal = "0.6"
 hyperlocal-windows = { git = "https://github.com/Azure/hyperlocal-windows" }
 mio = "0.6"
 mio-named-pipes = "0.1"
-mio-uds-windows = { git = "https://github.com/Azure/mio-uds-windows" }
+mio-uds-windows = { git = "https://github.com/Azure/mio-uds-windows.git" }
 miow = "0.3"



$ cargo update -p winapi:0.3.5 --precise 0.3.5

    Updating crates.io index
    Updating git repository `https://github.com/philipktlin/backtrace-rs`
    Updating git repository `https://github.com/philipktlin/cmake-rs`
    Updating git repository `https://github.com/philipktlin/dtoa`
    Updating git repository `https://github.com/philipktlin/iovec`
    Updating git repository `https://github.com/philipktlin/mio`
    Updating git repository `https://github.com/philipktlin/miow`
    Updating git repository `https://github.com/philipktlin/hjson-rust`
    Updating git repository `https://github.com/philipktlin/winapi-rs`
    Updating git repository `https://github.com/philipktlin/mio-uds-windows.git`
    Removing aho-corasick v0.5.3
      Adding backtrace v0.3.6 (https://github.com/philipktlin/backtrace-rs?branch=arm#a6caa2dd)
    Removing backtrace v0.3.6
      Adding backtrace-sys v0.1.16 (https://github.com/philipktlin/backtrace-rs?branch=arm#a6caa2dd)
    Removing backtrace-sys v0.1.16
      Adding cmake v0.1.30 (https://github.com/philipktlin/cmake-rs?branch=arm#886c0dd7)
    Removing cmake v0.1.30
      Adding dtoa v0.4.2 (https://github.com/philipktlin/dtoa?branch=arm#db97225f)
    Removing dtoa v0.4.2
      Adding iovec v0.1.2 (https://github.com/philipktlin/iovec?branch=arm#5d56185b)
    Removing iovec v0.1.2
    Removing kernel32-sys v0.2.2
    Removing memchr v0.1.11
      Adding mio v0.6.14 (https://github.com/philipktlin/mio?branch=arm#bd83233b)
    Removing mio v0.6.14
      Adding mio-uds-windows v0.1.0 (https://github.com/philipktlin/mio-uds-windows.git?branch=arm#581fcd3d)
    Removing mio-uds-windows v0.1.0 (https://github.com/Azure/mio-uds-windows.git#87a4a997)
      Adding miow v0.3.1 (https://github.com/philipktlin/miow?branch=arm#8f63c73f)
    Removing miow v0.2.1
    Removing miow v0.3.1
    Removing regex v0.1.80
    Removing regex-syntax v0.3.9
      Adding serde-hjson v0.8.1 (https://github.com/philipktlin/hjson-rust?branch=arm#e1fc93df)
    Removing serde-hjson v0.8.1
    Removing thread-id v2.0.0
    Removing thread_local v0.2.7
    Removing utf8-ranges v0.1.3
      Adding winapi v0.3.5 (https://github.com/philipktlin/winapi-rs?branch=arm/v0.3.5#12a51f5b)
    Removing winapi v0.2.8
    Removing winapi v0.3.5
    Removing winapi-build v0.1.1
      Adding winapi-i686-pc-windows-gnu v0.4.0 (https://github.com/philipktlin/winapi-rs?branch=arm/v0.3.5#12a51f5b)
    Removing winapi-i686-pc-windows-gnu v0.4.0
      Adding winapi-x86_64-pc-windows-gnu v0.4.0 (https://github.com/philipktlin/winapi-rs?branch=arm/v0.3.5#12a51f5b)
    Removing winapi-x86_64-pc-windows-gnu v0.4.0
    Removing ws2_32-sys v0.2.1



$ cargo update -p mio-uds-windows

    Updating git repository `https://github.com/philipktlin/mio-uds-windows.git`
    Updating crates.io index
    Updating git repository `https://github.com/Azure/mio-uds-windows.git`


$ cargo build --all

...
```

At this point, there is only one winapi and only one mio-uds-windows in Cargo.lock:

```

[[package]]
name = "mio-uds"
version = "0.6.7"
source = "registry+https://github.com/rust-lang/crates.io-index"
dependencies = [
 "iovec 0.1.2 (git+https://github.com/philipktlin/iovec?branch=arm)",
 "libc 0.2.42 (registry+https://github.com/rust-lang/crates.io-index)",
 "mio 0.6.14 (git+https://github.com/philipktlin/mio?branch=arm)",
]

[[package]]
name = "mio-uds-windows"
version = "0.1.0"
source = "git+https://github.com/philipktlin/mio-uds-windows.git?branch=arm#581fcd3d659ecfe7bf56f4e75e3402dca470c631"
dependencies = [
 "iovec 0.1.2 (git+https://github.com/philipktlin/iovec?branch=arm)",
 "lazycell 1.2.0 (registry+https://github.com/rust-lang/crates.io-index)",
 "log 0.4.5 (registry+https://github.com/rust-lang/crates.io-index)",
 "mio 0.6.14 (git+https://github.com/philipktlin/mio?branch=arm)",
 "miow 0.3.1 (git+https://github.com/philipktlin/miow?branch=arm)",
 "winapi 0.3.5 (git+https://github.com/philipktlin/winapi-rs?branch=arm/v0.3.5)",
]

[[package]]
name = "miow"
version = "0.3.1"
source = "git+https://github.com/philipktlin/miow?branch=arm#8f63c73f8983df08b66e2aa6d44a9790e91c94bb"
dependencies = [
 "socket2 0.3.5 (registry+https://github.com/rust-lang/crates.io-index)",
 "winapi 0.3.5 (git+https://github.com/philipktlin/winapi-rs?branch=arm/v0.3.5)",
]

...

[[package]]
name = "win-logger"
version = "0.1.0"
dependencies = [
 "edgelet-utils 0.1.0",
 "failure 0.1.2 (registry+https://github.com/rust-lang/crates.io-index)",
 "log 0.4.5 (registry+https://github.com/rust-lang/crates.io-index)",
 "winapi 0.3.5 (git+https://github.com/philipktlin/winapi-rs?branch=arm/v0.3.5)",
]

[[package]]
name = "winapi"
version = "0.3.5"
source = "git+https://github.com/philipktlin/winapi-rs?branch=arm/v0.3.5#12a51f5b56cffc81bd94eea7d6a8d006c9dec2f8"
dependencies = [
 "winapi-i686-pc-windows-gnu 0.4.0 (git+https://github.com/philipktlin/winapi-rs?branch=arm/v0.3.5)",
 "winapi-x86_64-pc-windows-gnu 0.4.0 (git+https://github.com/philipktlin/winapi-rs?branch=arm/v0.3.5)",
]

[[package]]
name = "winapi-i686-pc-windows-gnu"
version = "0.4.0"
source = "git+https://github.com/philipktlin/winapi-rs?branch=arm/v0.3.5#12a51f5b56cffc81bd94eea7d6a8d006c9dec2f8"
```

chandde replied:
OK, now I can repro what you suggested locally, pushed a fix and started a new build

arsing commented on edgelet/build/windows/build.ps1:
```diff
@@ -1,23 +1,80 @@
 # Copyright (c) Microsoft. All rights reserved.
```
File seems to have a mix of LF and CRLF. Please normalize to LF.

chandde replied:
Fixed. Just curious, why LF in build.ps1 and CRLF in package.ps1?

arsing replied:
It was originally CRLF but another commit accidentally changed it to LF. I don't particularly care about which one it is, just that it's consistent and doesn't change unnecessarily. So if the rest of the file is CRLF then keep it CRLF. If it's LF then keep it LF.

arsing commented on edgelet/build/windows/package.ps1:
```diff
@@ -1,8 +1,17 @@
-Param([Switch] $CreateTemplate, [Switch] $CreateCab, [Switch] $SkipInstallCerts)
+Param(
```
File has a mix of line endings. Please normalize to CRLF.

chandde replied:
fixed

arsing commented on edgelet/build/windows/build.ps1:
```diff
@@ -1,23 +1,80 @@
\ No newline at end of file
+}
+
+$ErrorActionPreference = 'Stop'
```
Add empty line at end of file.

chandde replied:
fixed

arsing commented on edgelet/build/windows/build.ps1:
```diff
@@ -1,23 +1,80 @@
+[patch."https://github.com/Azure/mio-uds-windows"]
+mio-uds-windows = { git = "https://github.com/philipktlin/mio-uds-windows", branch = "arm" }
+
+[patch."https://github.com/Azure/hyperlocal-windows"]
```
You shouldn't need to patch [`hyperlocal-windows`](https://github.com/philipktlin/hyperlocal-windows/compare/master..arm) and [`tokio-uds-windows`](https://github.com/philipktlin/tokio-uds-windows/compare/master..arm) since they don't have any changes except to remove the `.git` suffix from the URL, which is not significant.

chandde replied:
see response to the other comment above

arsing replied:
Same here. I wrote this comment after I tested it on my machine.

chandde replied:
OK, now I can repro what you suggested locally, pushed a fix and started a new build

arsing commented on edgelet/hsm-sys/azure-iot-hsm-c/CMakeLists.txt:
```diff
@@ -186,3 +186,8 @@ set (CPACK_RPM_PACKAGE_LICENSE "Proprietary")
+    if(WIN32)
+        target_link_libraries(iothsm shell32)
+    endif(WIN32)
+endif()
```
Add empty line at end of file.

chandde commented:
latest build is here https://msazure.visualstudio.com/One/_build/results?buildId=21663410

arsing commented:
As we'd discussed in the meeting, you also need to update https://github.com/Azure/iotedge/blob/master/edgelet/build/windows/build-diagnostics.ps1 that builds the iotedge-diagnostics binary + image, and https://github.com/Azure/iotedge/blob/master/builds/misc/images.yaml to call the script for Windows ARM (just remove `arm32v7: 'false'`).

chandde commented:
> As we'd discussed in the meeting, you also need to update https://github.com/Azure/iotedge/blob/master/edgelet/build/windows/build-diagnostics.ps1 that builds the iotedge-diagnostics binary + image, and https://github.com/Azure/iotedge/blob/master/builds/misc/images.yaml to call the script for Windows ARM (just remove `arm32v7: 'false'`).

I was trying to make that change in a different PR, (almost done), but if you think it's more appropriate to get them done in one PR, I'll push more changes then

chandde commented:
images.yaml build failed, checking...

arsing commented on edgelet/build/windows/util.ps1:
```diff
@@ -1,57 +1,165 @@
+
+    # $env:PATH = $clPath + ";" + $env:PATH
+
+    $clpath1 = "C:\Program Files (x86)\Microsoft Visual Studio\2017\BuildTools\VC\Tools\MSVC\14.16.27023\bin\Hostx64\x64;"
```
These paths might break if the VSTS VMs / build VMs update to a newer VS, no? Can you discover the path using the loop you commented out above?

Also, can you add them only if `cl.exe` doesn't already exist in `PATH` ? ie do the `Get-Command` test first.

arsing commented on edgelet/build/windows/openssl.ps1:
```diff
@@ -29,12 +35,14 @@ function Get-OpenSSL
-    $strawberryPerlPath = "$env:HOMEDRIVE\vcpkg\Downloads\strawberry-perl-5.24.1.1-32bit-portable.zip"
-    Invoke-WebRequest -Uri $strawberryPerlUri -OutFile $strawberryPerlPath
+    
+    # strawberry perl is now being downloaded as part vcpkg setup, no need to manually set up
```
We download it from our storage account because it has happened in the past where its download server went offline and broke our build. Please put this code back.

arsing commented on edgelet/build/windows/package.ps1:
```diff
@@ -54,8 +63,15 @@ Function New-Package([string] $Name, [string] $Version)
+    if($Arm)
+    {
+        # pkggen cannot find makecat.exe from below folder at runtime
+        $env:PATH = "C:\Program Files (x86)\Windows Kits\10\bin\10.0.17763.0\x64;" + $env:PATH
```
Don't hard-code this path.

arsing commented on edgelet/build/windows/util.ps1:
```diff
@@ -1,57 +1,216 @@
-        }
-    }
-}
+# Copyright (c) Microsoft. All rights reserved.
```
Seems you changed the line endings of this file. Revert it to whatever it was before.

arsing commented on edgelet/build/windows/util.ps1:
```diff
@@ -1,57 +1,216 @@
+    $ErrorActionPreference = 'Stop'
+}
+
+function ReplacePrivateRustInPath
```
```powershell
function ReplacePrivateRustInPath {
    Write-Host 'Remove cargo path in user profile from PATH, and add the private arm version to the PATH'

    $oldPath = $env:PATH

    [string[]] $newPaths = $env:PATH -split ';' |
        ?{
            $removePath = $_.Contains('.cargo')
            if ($removePath) {
                Write-Host "$_ is being removed from PATH"
            }
            -not $removePath
        }
    $newPaths += GetPrivateRustPath
    $env:PATH = $newPaths -join ';'

    $oldPath
}
```

arsing commented on edgelet/build/windows/util.ps1:
```diff
@@ -1,57 +1,216 @@
+    }
+
+    $edgefolder = Get-EdgeletFolder
+    Set-Location -Path $edgefolder
```
Don't change the current directory inside a function.

All cargo commands take a `--manifest-path` parameter, like you can see in other scripts. Use that instead.

chandde commented:
@arsing @philipktlin @myagley let me know if you have any other concerns, I'd happy to address.

arsing commented on edgelet/build/windows/package.ps1:
```diff
@@ -69,11 +101,15 @@ Function New-Package([string] $Name, [string] $Version)
-    $docker_cli_uri = 'https://mby.blob.core.windows.net/mby-win-amd64/docker-3.0.5.exe'
-    $docker_cli_license_uri = 'https://mby.blob.core.windows.net/mby/LICENSE-cli'
+    $docker_cli_uri = if($Arm) { 'https://edgebuild.blob.core.windows.net/iotedge-win-arm32v7-tools/docker.exe'} else { 'https://mby.blob.core.windows.net/mby-win-amd64/docker-3.0.5.exe' }
+    $docker_cli_license_uri = if ($Arm) { 'https://edgebuild.blob.core.windows.net/iotedge-win-arm32v7-tools/dockerd.exe' } else { 'https://mby.blob.core.windows.net/mby/LICENSE-cli' }
```
`dockerd.exe` URL is for `docker_engine_uri`, not `docker_cli_license_uri`. Bad merge?

arsing commented:
Between https://github.com/Azure/iotedge/commit/a793d0b271bd5ff3adb215cf124be7215cc6d51c#diff-d39d0000961ae60c470c48f100fc1e46R92 and https://github.com/Azure/iotedge/commit/c9e7bdf8780d64222aff4fbc3a7a8e6e52971e7f#diff-d39d0000961ae60c470c48f100fc1e46R121 the ARM package would've been broken because it would have had amd64 `dockerd.exe`. Please test with the fixed package.

chandde commented:
> Between [a793d0b#diff-d39d0000961ae60c470c48f100fc1e46R92](https://github.com/Azure/iotedge/commit/a793d0b271bd5ff3adb215cf124be7215cc6d51c#diff-d39d0000961ae60c470c48f100fc1e46R92) and [c9e7bdf#diff-d39d0000961ae60c470c48f100fc1e46R121](https://github.com/Azure/iotedge/commit/c9e7bdf8780d64222aff4fbc3a7a8e6e52971e7f#diff-d39d0000961ae60c470c48f100fc1e46R121) the ARM package would've been broken because it would have had amd64 `dockerd.exe`. Please test with the fixed package.

Kicking off two new builds for edgelet and images, will do e2e verification on the new builds.

arsing commented:
I already ran the builds for the last commit I pushed. You can see it in the list of checks.
